source/llassetgen/source/Image.cpp

```diff
@@ -0,0 +1,360 @@
+        png_voidp a = png_get_io_ptr(png);
+        ((std::istream*)a)->read((char*)data, length);
+    }
+
+    void Image::writeData(png_structp png, png_bytep data, png_size_t length) {
+        png_voidp a = png_get_io_ptr(png);
+        ((std::ostream*)a)->write((char*)data, length);
+    }
+
+    void Image::flushData(png_structp png) {
+        png_voidp a = png_get_io_ptr(png);
+        ((std::ostream*)a)->flush();
+    }
+
+    uint32_t Image::reduceBitDepth(uint32_t in, uint8_t in_bitDepth, uint8_t out_bitDepth) {
+        uint32_t in_max = (1 << in_bitDepth) - 1,
+                 out_max = (1 << out_bitDepth) - 1;
+        return in * out_max / in_max;
+    }
+
+    Image::Image(const std::string& filepath, uint8_t _bitDepth) {
+        std::ifstream in_file(filepath, std::ifstream::in | std::ifstream::binary);
+
+        png_byte pngsig[8];
+        in_file.read((char*)pngsig, 8);
+        if (!in_file.good()) {
+            std::cerr << "could not read from file";
+            abort();
+        }
+        if (png_sig_cmp(pngsig, 0, 8)) {
+            std::cerr << "no PNG file signature";
+            abort();
+        }
+
+        png_structp png = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+        if (!png) {
+            std::cerr << "failed to create png read struct";
+            abort();
+        }
+
+        png_infop info = png_create_info_struct(png);
+        if (!info) {
+            png_destroy_read_struct(&png, (png_infopp)0, (png_infopp)0);
+            std::cerr << "failed to create png info struct";
+            abort();
+        }
+
+        if (setjmp(png_jmpbuf(png))) {
+            std::cerr << "pnglib caused a longjump due to an error";
+            abort();
+        }
+
+        png_set_read_fn(png, (png_voidp)&in_file, readData);
+
+        png_set_sig_bytes(png, 8);
+
+        png_read_info(png, info);
+
+        min.x = 0;
+        min.y = 0;
+        max.x = png_get_image_width(png, info);
+        max.y = png_get_image_height(png, info);
+        uint32_t color_type = png_get_color_type(png, info);
+
+        if (color_type == PNG_COLOR_TYPE_GRAY) {
+            if (png_get_bit_depth(png, info) < 8) {
+                png_set_expand_gray_1_2_4_to_8(png);
+            }
+        } else {
+            if (png_get_valid(png, info, PNG_INFO_tRNS)) {
+                png_set_tRNS_to_alpha(png);
+            }
+            if (color_type == PNG_COLOR_TYPE_PALETTE) {
+                png_set_palette_to_rgb(png);
+            } else if (color_type == PNG_COLOR_TYPE_RGB ||
+                color_type == PNG_COLOR_TYPE_RGBA ||
+                color_type == PNG_COLOR_TYPE_RGB_ALPHA) {
+                png_set_rgb_to_gray_fixed(png, 1, -1, -1);
+            }
+            /* TODO: all color types:
+            PNG_COLOR_TYPE_GRAY_ALPHA
+            PNG_COLOR_MASK_PALETTE
+            PNG_COLOR_MASK_COLOR
+            PNG_COLOR_MASK_ALPHA*/
```

Comment:
Should probably be reported as an error somehow, instead of leaving the image in a uninitialized state.